File: src/main/java/net/heartshapedbox/math/two_d/LineSegment.java
package net.heartshapedbox.math.two_d;

import net.minecraft.util.math.Vec2f;

import java.util.Optional;

public class LineSegment {
    Vec2f a;
    Vec2f b;
    
    public LineSegment(Vec2f a, Vec2f b) {
        this.a = a;
        this.b = b;
    }
    
    public Optional<Vec2f> intersectFromLine(Line line, boolean checkY) {
        if (!checkY) {
            double aRes = line.solveForX(a.y);
            double bRes = line.solveForX(b.y);
            if (isNotValid(aRes) || isNotValid(bRes)) {
                return Optional.empty();
            }
            
            if (hasSignChange(a.x, aRes, bRes)) {
                Vec2f endA = new Vec2f((float)aRes, a.y);
                Vec2f endB = new Vec2f((float)bRes, b.y);
                
                double a1 = a.y - b.y;
                double b1 = b.x - a.x;
                double c1 = a1 * b.x + b1 * b.y;
    
                double a2 = endB.y - endA.y;
                double b2 = endA.x - endB.x;
                double c2 = a2 * endA.x + b2 * endA.y;
    
                double delta = a1 * b2 - a2 * b1;
                return Optional.of(new Vec2f((float) ((b2 * c1 - b1 * c2) / delta), (float) ((a1 * c2 - a2 * c1) / delta)));
            } else {
                return Optional.empty();
            }
        } else {
            double aRes = line.solveForY(a.x);
            double bRes = line.solveForY(b.x);
            if (isNotValid(aRes) || isNotValid(bRes)) {
                return Optional.empty();
            }
    
            if (hasSignChange(a.y, aRes, bRes)) {
                Vec2f endA = new Vec2f(a.x, (float)aRes);
                Vec2f endB = new Vec2f(b.x, (float)bRes);
        
                double a1 = a.y - b.y;
                double b1 = b.x - a.x;
                double c1 = a1 * b.x + b1 * b.y;
        
                double a2 = endB.y - endA.y;
                double b2 = endA.x - endB.x;
                double c2 = a2 * endA.x + b2 * endA.y;
        
                double delta = a1 * b2 - a2 * b1;
                return Optional.of(new Vec2f((float) ((b2 * c1 - b1 * c2) / delta), (float) ((a1 * c2 - a2 * c1) / delta)));
            } else {
                return Optional.empty();
            }
        }
    }
    
    private boolean hasSignChange(double base, double a, double b) {
        a -= base;
        b -= base;
        return (a >= 0 && b <= 0) || (a <= 0 && b >= 0);
    }
    
    private boolean isNotValid(double value) {
        return Double.isNaN(value) || Double.isInfinite(value);
    }
}

File: src/main/kotlin/io/github/faecraft/heartshapedbox/math/two_d/LineSegment.kt
package io.github.faecraft.heartshapedbox.math.two_d

import net.minecraft.util.math.Vec2f
import java.util.*

data class LineSegment(val a: Vec2f, val b: Vec2f) {
    fun intersectFromLine(line: Line, checkY: Boolean): Optional<Vec2f> {
        return if (!checkY) {
            val aRes = line.solveForX(a.y.toDouble())
            val bRes = line.solveForX(b.y.toDouble())
            if (isNotValid(aRes) || isNotValid(bRes)) {
                return Optional.empty()
            }
            if (hasSignChange(a.x.toDouble(), aRes, bRes)) {
                val endA = Vec2f(aRes.toFloat(), a.y)
                val endB = Vec2f(bRes.toFloat(), b.y)
                val a1 = (a.y - b.y).toDouble()
                val b1 = (b.x - a.x).toDouble()
                val c1 = a1 * b.x + b1 * b.y
                val a2 = (endB.y - endA.y).toDouble()
                val b2 = (endA.x - endB.x).toDouble()
                val c2 = a2 * endA.x + b2 * endA.y
                val delta = a1 * b2 - a2 * b1
                Optional.of(
                    Vec2f(
                        ((b2 * c1 - b1 * c2) / delta).toFloat(),
                        ((a1 * c2 - a2 * c1) / delta).toFloat()
                    )
                )
            } else {
                Optional.empty()
            }
        } else {
            val aRes = line.solveForY(a.x.toDouble())
            val bRes = line.solveForY(b.x.toDouble())
            if (isNotValid(aRes) || isNotValid(bRes)) {
                return Optional.empty()
            }
            if (hasSignChange(a.y.toDouble(), aRes, bRes)) {
                val endA = Vec2f(a.x, aRes.toFloat())
                val endB = Vec2f(b.x, bRes.toFloat())
                val a1 = (a.y - b.y).toDouble()
                val b1 = (b.x - a.x).toDouble()
                val c1 = a1 * b.x + b1 * b.y
                val a2 = (endB.y - endA.y).toDouble()
                val b2 = (endA.x - endB.x).toDouble()
                val c2 = a2 * endA.x + b2 * endA.y
                val delta = a1 * b2 - a2 * b1
                Optional.of(
                    Vec2f(
                        ((b2 * c1 - b1 * c2) / delta).toFloat(),
                        ((a1 * c2 - a2 * c1) / delta).toFloat()
                    )
                )
            } else {
                Optional.empty()
            }
        }
    }

    private fun hasSignChange(base: Double, a: Double, b: Double): Boolean {
        var mutA = a
        var mutB = b
        mutA -= base
        mutB -= base
        return mutA >= 0 && mutB <= 0 || mutA <= 0 && mutB >= 0
    }

    private fun isNotValid(value: Double): Boolean {
        return java.lang.Double.isNaN(value) || java.lang.Double.isInfinite(value)
    }
}
